### ISSUE TITLE
Inconsistent permutation spec between xla transpose and stablehlo::TransposeOp

### ISSUE THREAD

**Issue by anakinxc** [openxla/stablehlo#1339]
### What happened?

In [XLA transpose semantic](https://www.tensorflow.org/xla/operation_semantics#transpose) permutation means `∀ i . 0 ≤ i < rank ⇒ input_dimensions[permutation[i]] = output_dimensions[i]`, where [stablehlo spec](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#semantics-89) says `For all dimensions i in operand, dim(operand, i) = dim(result, j) where j = permutation[i]`. It seems permutation semantic is reversed between xla and stablehlo.

But when looking at code where import [xla into mhlo](https://github.com/openxla/xla/blob/f5e12ac6a66f511836f710a0b545508a48b84aaf/xla/translate/hlo_to_mhlo/hlo_function_importer.cc#L1589) and [mhlo to stablehlo](https://github.com/openxla/xla/blob/main/xla/mlir_hlo/mhlo/transforms/hlo_legalize_to_stablehlo/hlo_legalize_to_stablehlo.cc), also [test here](https://github.com/openxla/xla/blob/b96d38ae433c20ae7981f29ae52ff3c77ec52bf8/xla/translate/hlo_to_mhlo/tests/import.hlotxt#L1522), there seems no code that actually handles this spec difference.

Is this intended? Or it's a bug in stablehlo spec?

**Comment by sdasgup3** [comment]
Thanks @anakinxc for the thoughtful comment that, in general, a permutation `output  = permute(input, permutation)` when applied in wrong order `result[perm[i]] = input[i] vs result[i] = input[permutation[i]]` could produce wrong semantics. 

The above two implementations of permutation ultimately boils down to the relation between the input index, say `j = [j0, j1,...,jR-1]` and output 
index, say `i = [i0, i1, ..., iR1]`. Note that I have intentionally chosen `j` for input and `i` for output so as to match, respectively, the `operand` and `result` indices in the formalism of `transpose` op. 

According to above two implementations of of permutation, the relation between `i` and `j` could be 

1. `id = j[permutation[d]]` or 
2. `jd = i[permutation[id]]`. 


From the StableHLO semantics, we can see that it follows (1).
```
...  More Formally, result[i0, ..., iR-1] = operand[j0, ..., jR-1] where i[d] = j[permutation[d]].
```

We can informally read the above as 
for each index `i` in `result`, transpose op choose an index `j`  in `operand` such that result[i] will correspond to operand[j] and the relationship between `i` and `j` is as shown above. From XLA semantics, I can see, without reading too much into the exact index labels `i` vs `j` used for the formulation, that it is also following the same behavior

```
∀ i . 0 ≤ i < rank ⇒ input_dimensions[permutation[i]] = output_dimensions[i].
```

Let me know if it helps or you have further questions. Thank you for your feedback.

**Comment by anakinxc** [comment]
Hi @sdasgup3 

Let's use this one as an example:

```
// CHECK:  "mhlo.transpose"(%arg0) {permutation = dense<[0,2,3,1]> : tensor<4xi64>} : (tensor<1x1x4x6xi32>) -> tensor<1x4x6x1xi32>
ROOT %transpose.2 = s32[1,4,6,1] transpose(s32[1,1,4,6] %Arg_0.1), dimensions={0,2,3,1}
```

By xla semantic, `input_dims[permutation[i]] = output_dims[i]`,
which results to 
```
input_dim[0] = output_dim[0]
input_dim[2] = output_dim[1]
input_dim[3] = output_dim[2]
input_dim[1] = output_dim[3]
```
and output shape should be 1x4x6x1

But if we follows stablehlo semantic,` dim(operand, i) = dim(result, permutation[i])`
the result should be
```
input_dim[0] = output_dim[0]
input_dim[1] = output_dim[2]
input_dim[2] = output_dim[3]
input_dim[3] = output_dim[1]
```
and output shape becomes 1x6x1x4, which is not the same when one follows xla semenatic.

Let me know if I misinterpreted something here

Thanks

**Comment by anakinxc** [comment]
Hi @sdasgup3 

The one that you also feel not right is quoted from constraints C(3) of stablehlo::TransposeOp.

I agree the one in semantics is correct.

**Comment by sdasgup3** [comment]
Ah! I see. All the time I was referring to the semantics, but the question is regarding the dimension sizes. I agree (C3) does not seem right and should be fixed.

**Comment by sdasgup3** [comment]
I already submitted a fix :) 
Please have a look https://github.com/openxla/stablehlo/pull/1345

And again, thanks for the input!

**Comment by anakinxc** [comment]
> I already submitted a fix suggestion :) Please have a look #1345
> 
> And again, thanks for finding it out!

LGTM, cheers
